sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala

```diff
@@ -262,25 +262,39 @@ object AppendColumns {
   def apply[T : Encoder, U : Encoder](
       func: T => U,
       child: LogicalPlan): AppendColumns = {
+    val outputEncoder = encoderFor[U]
+    val namedExpressions = if (!outputEncoder.isSerializedAsStruct) {
+      assert(outputEncoder.namedExpressions.length == 1)
+      outputEncoder.namedExpressions.map(Alias(_, "key")())
+    } else {
+      outputEncoder.namedExpressions
+    }
     new AppendColumns(
       func.asInstanceOf[Any => Any],
       implicitly[Encoder[T]].clsTag.runtimeClass,
       implicitly[Encoder[T]].schema,
       UnresolvedDeserializer(encoderFor[T].deserializer),
-      encoderFor[U].namedExpressions,
+      namedExpressions,
       child)
   }
 
   def apply[T : Encoder, U : Encoder](
       func: T => U,
       inputAttributes: Seq[Attribute],
       child: LogicalPlan): AppendColumns = {
+    val outputEncoder = encoderFor[U]
+    val namedExpressions = if (!outputEncoder.isSerializedAsStruct) {
+      assert(outputEncoder.namedExpressions.length == 1)
+      outputEncoder.namedExpressions.map(Alias(_, "key")())
+    } else {
+      outputEncoder.namedExpressions
```

Comment:
so we may still fail if `T` and `U` are case classes and have conflict field names?

Reply:
yea, but for such cases, seems it is more complicated as we can't simply create aliases for serializer fields. Because in methods like `mapGroups`, we need to access original fields of key.

Reply:
if that is the case, I feel it better to ask users to resolve conflict manually, by adding alias.

Reply:
You mean we detect conflicting case, and show some error messages to ask users for that?

Reply:
We can improve the `CheckAnalysis` to detect this case, and improve the error message to ask users to do alias.

Reply:
I just tried to manually add alias. It seems not working as we expect:

```scala
val ds = Seq(Some(("a", 10)), Some(("a", 20)), Some(("b", 1)), Some(("b", 2)), Some(("c", 1)), None).toDS()

val newDS = ds.select(expr("value").as("opt").as[Option[(String, Int)]])

// schema
root                                                                                                                                                   
 |-- value: struct (nullable = true)                                                                                                                   
 |    |-- _1: string (nullable = true)                                                                                                          
 |    |-- _2: integer (nullable = false)

// physical plan
*(1) SerializeFromObject [if (isnull(unwrapoption(ObjectType(class scala.Tuple2), input[0, scala.Option, true]))) null else named_struct(_1, staticinvo
ke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(unwrapoption(ObjectType(class scala.Tuple2), input[0, scala.Op
tion, true]))._1, true, false), _2, assertnotnull(unwrapoption(ObjectType(class scala.Tuple2), input[0, scala.Option, true]))._2) AS value#5482]
+- *(1) MapElements <function1>, obj#5481: scala.Option                                             
   +- *(1) DeserializeToObject newInstance(class scala.Tuple1), obj#5480: scala.Tuple1                                  
      +- *(1) Project [value#5473 AS opt#5476]                                                                         
         +- LocalTableScan [value#5473]           
```

So even we add alias before `groupByKey`, it can't change dataset's output names.

Reply:
is this a special case of option of product? can you try pritimive type and product type?

Reply:
For primitive type and product type, looks like it works:
```scala
test("typed aggregation on primitive data") {
  val ds = Seq(1, 2, 3).toDS()
  val agg = ds.select(expr("value").as("data").as[Int])
    .groupByKey(_ >= 2)
    .agg(sum("data").as[Long], sum($"data" + 1).as[Long])
  agg.show()
}
```
```
+-----+---------+---------------+
|value|sum(data)|sum((data + 1))|
+-----+---------+---------------+
|false|        1|              2|
| true|        5|              7|
+-----+---------+---------------+
```

```scala
test("typed aggregation on product data") {
  val ds = Seq((1, 2), (2, 3), (3, 4)).toDS()
  val agg = ds.select(expr("_1").as("a").as[Int], expr("_2").as("b").as[Int])
    .groupByKey(_._1).agg(sum("a").as[Int], sum($"b" + 1).as[Int])
  agg.show
}
```
```
[info] - typed aggregation on primitive data (192 milliseconds)
+-----+------+------------+
|value|sum(a)|sum((b + 1))|
+-----+------+------------+
|    3|     3|           5|
|    1|     1|           3|
|    2|     2|           4|
+-----+------+------------+

```

Reply:
I tried to add check into `CheckAnalysis` to detect such `AppendColumns`. I found that we have many such use cases in `DatasetSuite`, e.g. `map and group by with class data`:

```scala
val ds: Dataset[(ClassData, Long)] = Seq(ClassData("one", 1), ClassData("two", 2)).toDS()
  .map(c => ClassData(c.a, c.b + 1))
  .groupByKey(p => p).count()
```

If users don't access original output like this patch shows, it won't cause problem. So I'm thinking if we disallow it at all, it is a behavior change. Should we only fail the `groupByKey` query accessing ambiguous field names? Or we should disallow at all if there is any conflicting name?

Reply:
> Should we only fail the groupByKey query accessing ambiguous field names?

Yes. When we have unresolved attributes, check if the child plan is `AppendColumns`.

Reply:
Actually such query won't reach `CheckAnalysis` because it will fail in resolving due to ambiguous references. So I think we should add such check in `Analyzer` when finding ambiguous references, instead of `CheckAnalysis`.

Reply:
ah i see. Then maybe we should just leave it instead of hacking the `AttributeSeq.resolve`.

Reply:
Can you elaborate it more?

Reply:
Oh, I see. Do you mean we don't do such check in `AttributeSeq.resolve` but leave it to `CheckAnalysis`?

Reply:
I mean, maybe we should just leave this problem. I'm not sure how hacky it is to detect the `AppendColumns` in this case. Maybe we can have more confidence if you have a PR ready.

Reply:
Thanks, I see. For this primitive type case, is current fix ok? Or we should deal with case classes together?

Reply:
I wouldn't special-case primitive type while this is a general problem.

Reply:
I wouldn't special-case primitive type while this is a general problem.

Reply:
Ok. I will try to make a PR and see if we can have better fix for this. Thanks for suggestion.